cdc/sinkv2/eventsink/mq/producer/kafka_producer.go

```diff
@@ -0,0 +1,243 @@
+	id model.ChangeFeedID
+	// role indicates this sink used for what.
+	role util.Role
+	// We hold the client to make close operation faster.
+	// Please see the comment of Close().
+	client sarama.Client
+	// asyncProducer is used to send messages to kafka asynchronously.
+	asyncProducer sarama.AsyncProducer
+	// closedMu is used to protect `closed`.
+	// We need to ensure that never write to closed producers.
+	closedMu sync.RWMutex
+	// closed is used to indicate whether the producer is closed.
+	// We also use it to guard against double closed.
+	closed bool
+	// closedChan is used to notify the run loop to exit.
+	closedChan chan struct{}
+	// failpointCh is used to notify the run loop to failpoint.
+	// Only used in test.
+	failpointCh chan error
+}
+
+func (k *kafkaProducer) AsyncSendMessage(
+	ctx context.Context, topic string,
+	partition int32, message *codec.MQMessage,
+) error {
+	// We have to hold the lock to prevent write to closed producer.
+	// Close may be blocked for a long time.
+	k.closedMu.RLock()
+	defer k.closedMu.RUnlock()
+
+	// If the producer is closed, we should skip the message.
+	if k.closed {
+		// We return the context error directly rather than other error.
+		// Because if producer already closed, it means the context maybe is canceling or canceled.
+		// So we shouldn't return other errors to the caller to mess up the shutdown process.
+		return ctx.Err()
+	}
+	failpoint.Inject("KafkaSinkAsyncSendError", func() {
+		// simulate sending message to input channel successfully but flushing
+		// message to Kafka meets error
+		log.Info("failpoint error injected", zap.String("namespace", k.id.Namespace),
+			zap.String("changefeed", k.id.ID), zap.Any("role", k.role))
+		k.failpointCh <- errors.New("kafka sink injected error")
+		failpoint.Return(nil)
+	})
+
+	failpoint.Inject("SinkFlushDMLPanic", func() {
+		time.Sleep(time.Second)
+		log.Panic("SinkFlushDMLPanic",
+			zap.String("namespace", k.id.Namespace),
+			zap.String("changefeed", k.id.ID), zap.Any("role", k.role))
+	})
+
+	msg := &sarama.ProducerMessage{
+		Topic:     topic,
+		Partition: partition,
+		Key:       sarama.StringEncoder(message.Key),
+		Value:     sarama.ByteEncoder(message.Value),
+		Metadata:  messageMetaData{callback: message.Callback},
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case k.asyncProducer.Input() <- msg:
+	}
+	return nil
+}
+
+func (k *kafkaProducer) Close() {
+	// We have to hold the lock to prevent write to closed producer.
+	k.closedMu.Lock()
+	defer k.closedMu.Unlock()
+	// If the producer was already closed, we should skip the close operation.
+	if k.closed {
+		// We need to guard against double closed the clients,
+		// which could lead to panic.
+		log.Warn("kafka producer already closed",
+			zap.String("namespace", k.id.Namespace),
+			zap.String("changefeed", k.id.ID),
+			zap.Any("role", k.role))
+		return
+	}
+	close(k.failpointCh)
+	// Notify the run loop to exit.
+	close(k.closedChan)
```

Comment:
This will make `run()` exit, meaning that nobody consumes from `k.asyncProducer.Successes()` and `<-k.asyncProducer.Errors()`. Will this be a problem? Is there a deadlock risk?

Reply:
I think it's OK.
Because close code looks like:
```go
func (p *asyncProducer) Close() error {
	p.AsyncClose()

	if p.conf.Producer.Return.Successes {
		go withRecover(func() {
			for range p.successes {
			}
		})
	}

	var errors ProducerErrors
	if p.conf.Producer.Return.Errors {
		for event := range p.errors {
			errors = append(errors, event)
		}
	} else {
		<-p.errors
	}

	if len(errors) > 0 {
		return errors
	}
	return nil
}
```

Reply:
There is a test case for this?

Reply:
Yes. TestProducerSendMsgFailed should cover it. I was able to get the error it returned in the previous test. Now it is not asserting this value because it is asynchronous.

I meaning I can get error from:
```go
if p.conf.Producer.Return.Errors {
		for event := range p.errors {
			errors = append(errors, event)
		}
	} else {
		<-p.errors
	}

```